src/gsHSplines/gsHDomain.h

```diff
@@ -440,7 +440,7 @@ class gsHDomain
     /// Returns the number of distinct knots in direction \a k of level \a lvl
     int numBreaks(int lvl, int k) const
     {
-        return (m_upperIndex[k] >> (m_indexLevel - lvl) );
+        return static_cast<typename util::make_unsigned<index_t>::type>(m_upperIndex[k]) >> (m_indexLevel - lvl);
```

Comment:
This cast is not good, the result is an int, but it seems that the return is cast to unsigned ?

Reply:
This cast is do the fact, that the left type (unsigned/signed) is decisive which shift (logical or arithmetical) is used (from compiler in assembler code, `shr` or `sar`. At this point we want a logical one, not a arithmetical one.

Reply:
No, in fact all the shifts here are arithmetic. When we shift right we want division, when left we want multiplication.

Also: the cast is not correct as well. If we needed a cast that would be 

static_cast<typename util::make_unsigned<T>::type>(m_upperIndex[k]) >> (m_indexLevel - lvl);

because m_indexLevel  is of type T.

Reply:
Your right with the cast - thought only of `T = index_t`
But original was `T = unsigned`, therefore I thought of the importance between a logical and a arithmetical shift. But, if you say it should be a shortcut to a division or multiplication, your only right in the unsigned case, or for positive numbers.
`x >> y` ==  `x / (1 << y)` only for unsigned and/or positive x.
If x can get negative, `>>` is not a correct division.

<details> <summary>unsigned x >> y</summary>

```cpp
int f1(unsigned x, int y) {
    return x >> y;
}
```
```ass
f1(unsigned int, int):
        mov     eax, edi
        mov     ecx, esi
        shr     eax, cl
        ret
```
```cpp
int f2(unsigned x, int y) {
    return x / (1 << y);
}
```
```ass
f2(unsigned int, int):
        mov     eax, edi
        mov     ecx, esi
        shr     eax, cl
        ret
```
</details>
<details> <summary>int x >> y</summary>

```cpp
int f1(int x, int y) {
    return x >> y;
}
```
```ass
f1(int, int):
        mov     eax, edi
        mov     ecx, esi
        sar     eax, cl
        ret
```
```cpp
int f2(int x, int y) {
    return x / (1 << y);
}
```
```ass
f2(int, int):
        mov     ecx, esi
        mov     eax, 1
        sal     eax, cl
        mov     ecx, eax
        mov     eax, edi
        cdq
        idiv    ecx
        ret
```
</details>

Example:
-1 / 2**1 = -1 or 0 ?

Reply:
Can it get negative?